File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.py
import sys
input = sys.stdin.readline
from collections import deque
# sys.stdin = open("input.txt", "r")

def bfs(start):
    q = deque([start])
    visited[start] = 1
    cnt = 1

    while q:
        cur = q.popleft()
        for next in edge[cur]:
            if visited[next] == 0:
                visited[next] = 1
                q.append(next)
                cnt += 1
    return cnt

maxTotal = 0
comList = []
n, m = map(int, input().split())
edge = [[] for _ in range(n+1)]
for _ in range(m):
    node1, node2 = map(int, input().split())
    edge[node2].append(node1)

for i in range(1, n+1):
    visited = [0 for _ in range(n + 1)]
    curTotal = bfs(i)
    if maxTotal < curTotal:
        maxTotal = curTotal
        comList = [i]
    elif maxTotal == curTotal:
        comList.append(i)
comList.sort()
print(*comList)

File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

vector<int> edge[100001];
bool visited[100001];

int bfs(int start, int n) {
    queue<int> q;
    q.push(start);
    visited[start] = true;
    int cnt = 1;
    
    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        
        for (int next : edge[cur]) {
            if (!visited[next]) {
                visited[next] = true;
                q.push(next);
                cnt++;
            }
        }
    }
    
    return cnt;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    
    int n, m;
    cin >> n >> m;
    
    for (int i = 0; i < m; i++) {
        int node1, node2;
        cin >> node1 >> node2;
        edge[node2].push_back(node1);
    }
    
    int maxTotal = 0;
    vector<int> comList;
    
    for (int i = 1; i <= n; i++) {
        fill(begin(visited), end(visited), false);
        int curTotal = bfs(i, n);
        
        if (maxTotal < curTotal) {
            maxTotal = curTotal;
            comList = {i};
        } else if (maxTotal == curTotal) {
            comList.push_back(i);
        }
    }
    
    for (int i : comList) {
        cout << i << " ";
    }
    cout << "\n";
    
    return 0;
}
